src/ZoneServer/World/Storage/Storage.cs

```diff
@@ -0,0 +1,691 @@
+			{
+				lock (_syncLock)
+				{
+					_storageItems.Remove(position);
+					removedAmount = item.Amount;
+				}
+			}
+			return StorageResult.Success;
+		}
+
+		/// <summary>
+		/// Removes an item from storage in a specific position.
+		/// Does not update client.
+		/// </summary>
+		/// <param name="item">Item to remove</param>
+		/// <param name="amount">Amount of the item</param>
+		/// <param name="removedAmount">Amount that was removed. Defaults to 0</param>
+		/// <returns></returns>
+		public StorageResult RemoveAtPosition(int position, int amount, out int removedAmount)
+		{
+			removedAmount = 0;
+
+			if (amount <= 0)
+				return StorageResult.InvalidOperation;
+
+			var result = _storageItems.TryGetValue(position, out var item);
+			if (!result)
+				return StorageResult.ItemNotFound;
+
+			return Remove(item, amount, out var removedPosition, out removedAmount);
+		}
+
+		/// <summary>
+		/// Returns true if storage is full.
+		/// </summary>
+		/// <returns></returns>
+		public bool CheckStorageFull()
+		{
+			lock (_syncLock)
+				return _storageItems.Count() >= _maxStorage;
+		}
+
+		/// <summary>
+		/// Returns a dictionary of items in each position
+		/// of storage.
+		/// </summary>
+		/// <returns></returns>
+		public Dictionary<int, Item> GetStorage()
+		{
+			lock (_syncLock)
+				return _storageItems.ToDictionary(kvp => kvp.Key, kvp => kvp.Value);
+		}
+
+		/// <summary>
+		/// Returns specific item, or null if it doesn't exist.
+		/// </summary>
+		/// <param name="objectId">Item object id</param>
+		/// <param name="item">Item found</param>
+		/// <param name="position">Position of item found in storage</param>
+		/// <returns></returns>
+		public StorageResult TryGetItem(long objectId, out Item item, out int position)
+		{
+			item = null;
+			position = -1;
+
+			lock (_syncLock)
+			{
+				var storageItem = _storageItems.FirstOrDefault(v => v.Value.ObjectId == objectId);
+				if (storageItem.Value != null)
+				{
+					item = storageItem.Value;
+					position = storageItem.Key;
+					return StorageResult.Success;
+				}
+			}
+
+			return StorageResult.ItemNotFound;
+		}
+
+		/// <summary>
+		/// Returns specific item in position, or null
+		/// if it doesn't exist.
+		/// </summary>
+		/// <param name="position"></param>
+		/// <returns></returns>
+		public Item TryGetItemAtPosition(int position)
+		{
+			lock (_syncLock)
+				if (_storageItems.TryGetValue(position, out var item))
+					return item;
+
+			return null;
+		}
+
+		/// <summary>
+		/// Returns how many items exist in storage
+		/// </summary>
+		/// <returns></returns>
+		public int GetItemCount()
+		{
+			lock (_syncLock)
+				return _storageItems.Count();
+		}
+
+		/// <summary>
+		/// Returns the max size of this storage
+		/// </summary>
+		/// <returns></returns>
+		public int GetStorageSize()
+		{
+			return _maxStorage;
+		}
+
+		/// <summary>
+		/// Increases max storage size by the given number.
+		/// Storages can never decrease in size.
```

Comment:
This, too, seems somewhat restrictive and also potentially problematic. What if someone accidentally called this method twice for example, should they then be forced to modify the database directly or attempt to understand the inner workings of this class? Don't you think the storage itself should be capable of decreasing the size, even if it might come at the cost of items? Though we could also simply rearrange the items or pass any potentially removed items to the caller in that case and let them deal with it. For example, they could be added to the inventory or be dropped on the floor.